rfcs/20201105-attribute-num-threads/README.md

```diff
@@ -0,0 +1,272 @@
+calling thread.
+
+By assigning `1` or higher value in the API, the user will instruct the library
+the upper limit for the number of threads for decomposition, and, during
+execution, the parallel sections will be executed by the threads in the
+threadpool.
+
+In the context of primitive cache, the number of threads passed by primitive
+attributes will be used to construct the cache key. In means that two primitives
+created using different `max_num_threads` will both be stored in the cache as
+separate implementations.
+
+Since the implementation of this RFC can target oneDNN v2.1 at the least, it is
+proposed to introduce the new API into `dnnl_threadpool.hpp`, where
+Threadpool-specific API is introduced.
+
+Pros:
+- Ability to have different problem decompositions for different primitives;
+- Flexibility with respect to user cost model and Threadpool implementation;
+- Minimal changes required in the library;
+
+Cons:
+- Upper limit for number of threads not applicable to other threading runtimes,
+  may cause some inconsistencies in the library code.
+
+### Option 2: new primitive attributes for all CPU threading runtimes
+
+In order to maintain consistency between threading runtimes, this option
+proposes to introduce a generic primitive attribute, that instructs the library
+the upper limit for thread count during primitive (descriptor) creation.
+
+<details>
+<summary>C API</summary>
+
+~~~cpp
+dnnl_primitive_attr_get_max_num_threads(const_dnnl_primitive_attr_t attr, uint32_t *max_num_threads);
+
+dnnl_primitive_attr_set_max_num_threads(dnnl_primitive_attr_t attr, uint32_t max_num_threads);
+~~~
+
+</details>
+
+<details>
+<summary>C++ API</summary>
+
+~~~cpp
+struct primitive_attr : public handle<dnnl_primitive_attr_t> {
+    ...
+
+    uint32_t get_max_num_threads() const {}
+
+    void set_max_num_threads(uint32_t nthr) {}
+
+    ...
+};
+~~~
+
+</details>
+
+The value provided for the primitive attribute will be used as an upper limit
+for number of threads for problem decomposition at primitive (descriptor)
+creation.
+
+The behavior of the primitive attribute will be identical to what is described
+in [Option 1](#option-1-new-api-in-dnnl::threadpool_interop-namespace), however,
+as mentioned above, it will be applicable for all threading runtimes.
+
+Pros:
+- Ability to have different problem decompositions for different primitives;
+- Flexibility with respect to user cost model and Threadpool implementation;
+- Minimal changes required in the library;
+- Consistency in the code for all threading runtimes;
+
+Cons:
+- Adjusting the number of threads for OMP and TBB may not yield any performance
+  benefits, making the API potentially redundant.
+
+### Option 3: thread decomposition at execution time
+
+For this option, use of `dnnl_get_max_threads()` inside primitive (descriptor)
+constructors will be restricted for some or all primitives, thereby performing
+the primitive decomposition at execution stage.
```

Comment:
Do you know how it would limit our jitting capabilities? For example, do we have jitter parameters that depend on the number of threads used?
I would think that the more we thread, the smaller are our block sizes, and these block sizes could be parameters to the code generators.

Reply:
As far as I know, blocking in oneDNN is mostly done for backward propagation with weights updates (and, occasionally, for forward propagation too). In the context of this RFC, where the goal is to improve inference performance, this is not such a major issue.

The bigger concern are the implications that the run time primitive decomposition may cause on the our programming model. There is some blocking even for forward propagation to improve performance. So, by improving threadpool performance issues with execution time decomposition may cause issues for OpenMP threading.